src/target/riscv/riscv.c

```diff
@@ -604,90 +606,349 @@ static int maybe_add_trigger_t1(struct target *target, struct trigger *trigger)
+	tdata1 = set_field(tdata1, CSR_MCONTROL6_MATCH, CSR_MCONTROL6_MATCH_NAPOT);
+	return tdata1;
+}
+
+static riscv_reg_t set_match_lt_t2(void)
+{
+	riscv_reg_t tdata1 = 0;
+	tdata1 = set_field(tdata1, CSR_MCONTROL_MATCH, CSR_MCONTROL_MATCH_LT);
+	return tdata1;
+}
+
+static riscv_reg_t set_match_lt_t6(void)
+{
+	riscv_reg_t tdata1 = 0;
+	tdata1 = set_field(tdata1, CSR_MCONTROL6_MATCH, CSR_MCONTROL6_MATCH_LT);
+	return tdata1;
+}
+
+static riscv_reg_t set_match_ge_t2(void)
+{
+	riscv_reg_t tdata1 = 0;
+	tdata1 = set_field(tdata1, CSR_MCONTROL_MATCH, CSR_MCONTROL_MATCH_GE);
+	return tdata1;
+}
+
+static riscv_reg_t set_match_ge_t6(void)
+{
+	riscv_reg_t tdata1 = 0;
+	tdata1 = set_field(tdata1, CSR_MCONTROL6_MATCH, CSR_MCONTROL6_MATCH_GE);
+	return tdata1;
+}
+
+static riscv_reg_t set_match_eq_t2(void)
+{
+	riscv_reg_t tdata1 = 0;
 	tdata1 = set_field(tdata1, CSR_MCONTROL_MATCH, CSR_MCONTROL_MATCH_EQUAL);
-	if (trigger->length == 1) {
-		tdata1 = set_field(tdata1, CSR_MCONTROL_SIZELO, CSR_MCONTROL_SIZELO_8BIT & 3);
-		tdata1 = set_field(tdata1, CSR_MCONTROL_SIZEHI, (CSR_MCONTROL_SIZELO_8BIT >> 2) & 3);
-	}
-	tdata1 = set_field(tdata1, CSR_MCONTROL_CHAIN, CSR_MCONTROL_CHAIN_DISABLED);
-	tdata2 = trigger->address;
-	ret = set_trigger(target, idx, tdata1, tdata2, CSR_MCONTROL_MASKMAX(riscv_xlen(target)));
-	if (ret != ERROR_OK)
-		return ret;
-	r->trigger_unique_id[idx] = trigger->unique_id;
-	return ERROR_OK;
+	return tdata1;
+}
+
+static riscv_reg_t set_match_eq_t6(void)
+{
+	riscv_reg_t tdata1 = 0;
+	tdata1 = set_field(tdata1, CSR_MCONTROL6_MATCH, CSR_MCONTROL6_MATCH_EQUAL);
+	return tdata1;
+}
+
+static struct match_triggers_field_info fill_match_triggers_field_info_t2(
+		struct target *target, struct trigger *trigger)
+{
+	struct match_triggers_field_info result = {
+		.common = set_common_t2(target, trigger),
+		.size = {
+			.any = set_size_any_t2(),
+			.s8bit = set_size_s8bit_t2()
+		},
+		.chain = {
+			.enable = set_chain_enable_t2(),
+			.disable = set_chain_disable_t2()
+		},
+		.match = {
+			.napot = set_match_napot_t2(),
+			.lt = set_match_lt_t2(),
+			.ge = set_match_ge_t2(),
+			.eq = set_match_eq_t2()
+		},
+		.tdata1_ignore_mask = CSR_MCONTROL_MASKMAX(riscv_xlen(target))
+	};
+	return result;
+}
+
+static struct match_triggers_field_info fill_match_triggers_field_info_t6(
+		struct target *target, struct trigger *trigger)
+{
+	struct match_triggers_field_info result = {
+		.common = set_common_t6(target, trigger),
+		.size = {
+			.any = set_size_any_t6(),
+			.s8bit = set_size_s8bit_t6()
+		},
+		.chain = {
+			.enable = set_chain_enable_t6(),
+			.disable = set_chain_disable_t6()
+		},
+		.match = {
+			.napot = set_match_napot_t6(),
+			.lt = set_match_lt_t6(),
+			.ge = set_match_ge_t6(),
+			.eq = set_match_eq_t6()
+		},
+		.tdata1_ignore_mask = 0
+	};
+	return result;
+}
+
+static int maybe_add_trigger_t2_t6(struct target *target,
+		struct trigger *trigger, struct match_triggers_field_info set)
+{
+	int ret = ERROR_OK;
```

Comment:
Could you please simplify it further by passing just the trigger type:

```suggestion
static int maybe_add_trigger_t2_t6(struct target *target,
		struct trigger *trigger, unsigned int trigger_type)
{
	assert(trigger_type == 2 || trigger_type == 6);
	int ret = ERROR_OK;
```

The `trigger_type` argument can be an unsigned int or an enum.